### PR TITLE
[BCG][3/N] Enable bcg on dsa & deepep a2a backend

### PR DESCRIPTION
**Pull request by Oasis-Git**
## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #30593427051](https://github.com/sgl-project/sglang/actions/runs/30593427051)
Latest PR Test (Extra): :white_check_mark: [Run #30593426887](https://github.com/sgl-project/sglang/actions/runs/30593426887)

### CODE REVIEW

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 1]
```diff
@@ -46,6 +52,67 @@
...
+
+    Mutates ``output`` (allocated by the caller inside the captured region,
+    i.e. graph-pool storage the pool reuses across shapes) and returns None,
+    so eager_on_graph retains no per-shape bridge tensor for this break.
+    """
+    from sglang.srt.layers.dp_attention import (
+        get_is_extend_in_batch,
+        set_is_extend_in_batch,
+    )
+    from sglang.srt.layers.moe.topk import StandardTopKOutput
```
Comment (merrymercy) [inline python/sglang/srt/layers/moe/ep_moe/layer.py]: put the imports to the top whenever possible

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 2]
```diff
@@ -46,6 +52,67 @@
...
+    hidden_states: torch.Tensor,
+    topk_weights: torch.Tensor,
+    topk_ids: torch.Tensor,
+    router_logits: torch.Tensor,
+    output: torch.Tensor,
+) -> None:
+    """Run the whole a2a MoE (dispatch -> experts -> combine) eagerly
+    between BCG segments, for any model using an a2a dispatcher.
+
+    The prefill graph runner pins is_extend_in_batch=False so MoE captured
```
Comment (merrymercy) [inline python/sglang/srt/layers/moe/ep_moe/layer.py]: why? then the prefill graph runner should set is_extend_in_batch=True.

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 3]
```diff
@@ -160,6 +227,20 @@ def forward(
         hidden_states: torch.Tensor,
         topk_output: TopKOutput,
     ):
+        if is_in_breakable_cuda_graph():
+            assert TopKOutputChecker.format_is_standard(
+                topk_output
+            ), "Only standard topk output is supported for breakable cuda graph"
+            output = torch.empty_like(hidden_states)
+            bcg_a2a_moe_forward(
```
Comment (merrymercy) [inline python/sglang/srt/layers/moe/ep_moe/layer.py]: it is a method of this class, why do you need to define it outside?
can the `eager_on_graph` work on class method?

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 4]
```diff
@@ -46,6 +52,67 @@
...
+) -> None:
+    # Capture-pass stand-in: zero the bridge buffer instead of running the
+    # real dispatch -> experts -> combine. The capture pass discards break
+    # outputs (only the buffer address is recorded), while the real body
+    # performs a rank-coupled NORMAL a2a whose per-break rendezvous during
+    # capture amplifies rank jitter and risks DeepEP's 100s CPU timeout.
+    output.zero_()
+
+
+_a2a_moe_forward_eager._capture_stub = _a2a_moe_forward_capture_stub
```
Comment (merrymercy) [inline python/sglang/srt/layers/moe/ep_moe/layer.py]: is this function ever used?

File: python/sglang/srt/managers/scheduler.py [hunk 5]
```diff
@@ -1789,8 +1789,15 @@ def init_dp_attn_adapter(self) -> None:
...
             spec_algorithm=self.spec_algorithm,
             get_require_mlp_sync=lambda: self.require_mlp_sync,
+            local_breakable_eligible_fn=self._local_breakable_replay_eligible,
         )
 
+    def _local_breakable_replay_eligible(self, batch) -> bool:
+        runner = getattr(self.tp_worker, "model_runner", None)
+        prefill_runner = getattr(runner, "prefill_cuda_graph_runner", None)
+        eligible = getattr(prefill_runner, "schedule_batch_replay_eligible", None)
+        return True if eligible is None else eligible(batch)
```
Comment (merrymercy) [inline python/sglang/srt/managers/scheduler.py]: getattr is not allowed in any new code https://github.com/sgl-project/sglang/blob/main/.claude/rules/no-getattr-defensive.md

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py [hunk 6]
```diff
@@ -241,6 +248,12 @@ def __init__(self, model_runner: ModelRunner):
             source=self.buffers,
         )
 
+        from sglang.srt.configs.model_config import is_deepseek_dsa
```
Comment (merrymercy) [inline python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: move imports to the top

File: python/sglang/srt/model_executor/runner_backend_utils/breakable_cuda_graph/breakable_cuda_graph.py [hunk 7]
```diff
@@ -231,9 +231,30 @@ def wrapper(*args, **kwargs):
...
+            # Segment teardown (capture_end + instantiate, worst case an
+            # allocator stall under memory pressure) is the slow, per-rank
+            # variable step of a capture. Re-sync ranks here — in a patient
+            # distributed barrier — before eager break fns that may contain
+            # rank-coupled collectives with short hard timeouts (DeepEP
+            # NORMAL dispatch spins at most LEGACY_NUM_CPU_TIMEOUT_SECS=100s,
+            # a compile-time constant). Capture-only by construction: replay
+            # bypasses this wrapper via replay_fn.
+            if capture._barrier_fn is not None:
+                capture._barrier_fn()
```
Comment (merrymercy) [inline python/sglang/srt/model_executor/runner_backend_utils/breakable_cuda_graph/breakable_cuda_graph.py:239]: do we really need this?

File: python/sglang/srt/models/deepseek_v2.py [hunk 8]
```diff
@@ -1210,7 +1210,12 @@ def forward_deepep(
             # router_logits: (num_tokens, n_experts)
             router_logits = self.gate(hidden_states, forward_batch=forward_batch)
             if not sbo_enabled_flag and self.num_fused_shared_experts == 0:
-                if self.alt_stream is not None:
+                # No gain under BCG: the a2a window worth hiding under runs
```
Comment (merrymercy) [inline python/sglang/srt/models/deepseek_v2.py]: does not make sense

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py [hunk 9]
```diff
@@ -678,6 +692,36 @@ def _restore_mha_capture_state(forward_batch: ForwardBatch) -> None:
         forward_batch.mha_return_lse = False
         forward_batch.set_attn_attend_prefix_cache(False)
 
+    def schedule_batch_replay_eligible(self, batch) -> bool:
```
Comment (merrymercy) [inline python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: we already have so many checks, why do we need a new one?

there is `can_run_graph` in python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py, `can_run_breakable_cuda_graph` in python/sglang/srt/managers/scheduler_components/dp_attn.py,

and now we have a new function here, let us unify all of these conditions in a single place

File: python/sglang/srt/layers/moe/ep_moe/layer.py (hunk 1 above)
Comment (Oasis-Git) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: resolved.

File: python/sglang/srt/layers/moe/ep_moe/layer.py (hunk 4 above)
Comment (Oasis-Git) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: This function is only used for the construction of bcg_function

File: python/sglang/srt/managers/scheduler.py (hunk 5 above)
Comment (Oasis-Git) [reply python/sglang/srt/managers/scheduler.py]: resolved/

File: python/sglang/srt/models/deepseek_v2.py (hunk 8 above)
Comment (Oasis-Git) [reply python/sglang/srt/models/deepseek_v2.py]: The overlap of dual stream may mainly work in decode, told from baizhou side. Also the two stream fork/join happens in the different segmentation which makes bcg hard to be compatible.

File: python/sglang/srt/layers/moe/ep_moe/layer.py (hunk 2 above)
Comment (Oasis-Git) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: This is related to the original implementation, especially about tc piecewise cuda graph side. It should be changed.

File: python/sglang/srt/model_executor/runner_backend_utils/breakable_cuda_graph/breakable_cuda_graph.py (hunk 7 above)
Comment (Oasis-Git) [reply python/sglang/srt/model_executor/runner_backend_utils/breakable_cuda_graph/breakable_cuda_graph.py:239]: I think a sync at the end of capture is beneficial since some of the ops with communication  may have timeout mechanism.

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py (hunk 9 above)
Comment (Oasis-Git) [reply python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: solved

File: python/sglang/srt/layers/moe/ep_moe/layer.py (hunk 4 above)
Comment (Oasis-Git) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: solved

File: python/sglang/srt/layers/moe/ep_moe/layer.py (hunk 3 above)
Comment (Oasis-Git) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: solved

**Comment by b8zhong** [comment]
Looks related https://github.com/sgl-project/sglang/actions/runs/30080250490/job/89440130032?pr=31987

**Comment by Oasis-Git** [comment]
CI Passed: https://github.com/sgl-project/sglang/actions/runs/30593426887

**Comment by Oasis-Git** [comment]
https://github.com/sgl-project/sglang/actions/runs/30593427051
